### PR TITLE
Allow unindent without selection

### LINKED ISSUE #50279: [Text Editor] Shift+Tab shortcut only works when multiple lines are selected
### Godot version

v4.0.dev.custom_build [cb4e42155]

### System information

manjaro Gnome 21, AMD R7 200 series, Intel i3 3.0 GHZ,8GB RAM

### Issue description

When you select multiple lines and press `SHIFT+TAB` the tab goes back one step which is ok:
before:
image
after:
image

 But if the cursor is over one line only (without selecting) `SHIFT+TAB` doesnt work:
before:
image

after:
image

PS: it only works if the single line is selected, but without any selection it doesnt work!

### Steps to reproduce

* Select Multiple lines
* press the shortcut
* put the cursor at a single line without selection
* press the shortcut

### Minimal reproduction project

N/a

### LINKED ISSUE #68154: Executing `CodeEdit.do_unindent` crashes Godot
### Godot version

4.0.beta.custom_build. 0a0e99cad

### System information

Ubuntu 22.04 - Nvidia GTX 970, Gnome shell 42 X11

### Issue description
```
extends Node
func _process(delta):

	var temp_variable382 = CodeEdit.new()
	add_child(temp_variable382)
	temp_variable382.set_indent_using_spaces(true)
	temp_variable382.indent_lines()
	temp_variable382.add_caret(-24, -19)
	temp_variable382.do_unindent()

```
Godot crashes with this backtrace

```
ERROR: Index p_index = 3 is out of bounds (size() = 0).
   at: get (./core/templates/cowdata.h:155)

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.0.beta.custom_build (0a0e99cadc6e250ff2ee27ef31760eb9c2e21040)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7fcc9ac2a520] (??:0)
[2] CowData<char32_t>::get(int) const (/home/runner/work/GodotBuilds/GodotBuilds/godot/./core/templates/cowdata.h:155 (discriminator 7))
[3] String::operator[](int) const (/home/runner/work/GodotBuilds/GodotBuilds/godot/./core/string/ustring.h:223)
[4] CodeEdit::do_unindent() (/home/runner/work/GodotBuilds/GodotBuilds/godot/scene/gui/code_edit.cpp:903 (discriminator 1))
[5] void call_with_variant_args_helper<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, Callable::CallError&, IndexSequence<>) (/home/runner/work/GodotBuilds/GodotBuilds/godot/./core/variant/binder_common.h:267 (discriminator 4))
[6] void call_with_variant_args_dv<__UnexistingClass>(__UnexistingClass*, void (__UnexistingClass::*)(), Variant const**, int, Callable::CallError&, Vector<Variant> const&) (/home/runner/work/GodotBuilds/GodotBuilds/godot/./core/variant/binder_common.h:410)
[7] MethodBindT<>::call(Object*, Variant const**, int, Callable::CallError&) (/home/runner/work/GodotBuilds/GodotBuilds/godot/./core/object/method_bind.h:320)
[8] Object::callp(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/GodotBuilds/GodotBuilds/godot/core/object/object.cpp:733 (discriminator 1))
[9] Variant::callp(StringName const&, Variant const**, int, Variant&, Callable::CallError&) (/home/runner/work/GodotBuilds/GodotBuilds/godot/core/variant/variant_call.cpp:1048)
[10] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/home/runner/work/GodotBuilds/GodotBuilds/godot/modules/gdscript/gdscript_vm.cpp:1556)
[11] GDScriptInstance::callp(StringName const&, Variant const**, int, Callable::CallError&) (/home/runner/work/GodotBuilds/GodotBuilds/godot/modules/gdscript/gdscript.cpp:1634)
[12] bool Node::_gdvirtual__process_call<false>(double) (/home/runner/work/GodotBuilds/GodotBuilds/godot/scene/main/node.h:238 (discriminator 5))
[13] Node::_notification(int) (/home/runner/work/GodotBuilds/GodotBuilds/godot/scene/main/node.cpp:57)
[14] Node::_notificationv(int, bool) (/home/runner/work/GodotBuilds/GodotBuilds/godot/./scene/main/node.h:45 (discriminator 14))
[15] Object::notification(int, bool) (/home/runner/work/GodotBuilds/GodotBuilds/godot/core/object/object.cpp:792)
[16] SceneTree::_notify_group_pause(StringName const&, int) (/home/runner/work/GodotBuilds/GodotBuilds/godot/scene/main/scene_tree.cpp:868)
[17] SceneTree::process(double) (/home/runner/work/GodotBuilds/GodotBuilds/godot/scene/main/scene_tree.cpp:468)
[18] Main::iteration() (/home/runner/work/GodotBuilds/GodotBuilds/godot/main/main.cpp:3170)
[19] OS_LinuxBSD::run() (/home/runner/work/GodotBuilds/GodotBuilds/godot/platform/linuxbsd/os_linuxbsd.cpp:776)
[20] godot4(main+0x19f) [0x2641a15] (/home/runner/work/GodotBuilds/GodotBuilds/godot/platform/linuxbsd/godot_linuxbsd.cpp:75)
[21] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fcc9ac11d90] (??:0)
[22] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7fcc9ac11e40] (??:0)
[23] godot4(_start+0x25) [0x26417b5] (??:?)
-- END OF BACKTRACE --
================================================================

```

This example was found by Godot fuzzer - [Qarminer](https://github.com/qarmin/Qarminer), so it is quite unlikelly that this code could be used in real project, but still this should be handled gracefully.

### Steps to reproduce

Above

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by KoBeWi**
Revival of #50289
Fixes #50279
Fixes #68154

### CODE REVIEW

**Comment by KoBeWi** [comment]
This seems to fail 2 test cases.
```
test[tab]
```
is expected to become
```
test
```
after unindent (i.e. whitespace to the right is removed). This behavior is probably what caused the original issue. Any reason why it would work like that? Is it common in code editors?

**Comment by KoBeWi** [comment]
Rebased. I also changed the failing test. tbh the two behaviors aren't really conflicting (i.e. Shift+Tab can remove trailing tab and unindent if there isn't one), but I'm just reviving a PR here, it's not like I know what it does >_>

@Paulb23 Review would be appreciated, the issue gets in my way too often.

**Comment by MewPurPur** [comment]
Makes sense to me. I don't see why the current implementation would be desired - if I need to remove whitespace, I can just press Backspace. Removing previous whitespace, instead of just tabs at the beginning of a line, idk about that but seems pointless.

File: scene/gui/code_edit.cpp [hunk 1]
```diff
@@ -2180,7 +2136,7 @@ void CodeEdit::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_auto_indent_prefixes"), &CodeEdit::get_auto_indent_prefixes);
 
 	ClassDB::bind_method(D_METHOD("do_indent"), &CodeEdit::do_indent);
-	ClassDB::bind_method(D_METHOD("do_unindent"), &CodeEdit::do_unindent);
+	ClassDB::bind_method(D_METHOD("do_unindent"), &CodeEdit::unindent_lines);
 
 	ClassDB::bind_method(D_METHOD("indent_lines"), &CodeEdit::indent_lines);
 	ClassDB::bind_method(D_METHOD("unindent_lines"), &CodeEdit::unindent_lines);
```
Comment (akien-mga) [inline scene/gui/code_edit.cpp:2142]: What's the point of having both `do_unident` and `unident_lines` exposed but duplicates?

Comment (KoBeWi) [reply scene/gui/code_edit.cpp:2142]: Eh, it's an oversight. I don't know how to handle this tbh.
Maybe I should restore `do_unindent()`, but keep `ui_text_dedent` as `unindent_lines()`? I don't even know what is `do_unindent()` for, its behavior is so useless >_>

File: doc/classes/CodeEdit.xml [hunk 2]
```diff
@@ -135,7 +135,7 @@
 		<method name="do_unindent">
 			<return type="void" />
 			<description>
-				Perform an unindent as if the user activated the "ui_text_unindent" action.
+				Unindents line.
```
Comment (KoBeWi) [inline doc/classes/CodeEdit.xml:138]: Needs better description (it already did tbh).

**Comment by KoBeWi** [comment]
I reworked the PR. It no longer removes `do_unindent()`, it simply changes the method used for dedent action.

**Comment by Paulb23** [comment]
I would continue to remove `do_unindent` then I think we can merge this for 4.0, as this would fix #68154. Otherwise it requires a bit of a rework for multi-carets. Can add it back later if needed.

**Comment by akien-mga** [comment]
What's the difference between the remaining `do_indent` and `indent_lines`?

**Comment by Paulb23** [comment]
Some of those test are probably still worth keeping.

> What's the difference between the remaining `do_indent` and `indent_lines`?

`do_indent` is called when the user presses tab, so will insert a tab/space at the caret(s). However, if there is a selection it will call `indent_lines`, which will move / indent lines that have a caret or selection. Similar to how the JetBrains IDE's work.

**Comment by KoBeWi** [comment]
> Some of those test are probably still worth keeping.

So should I change them to `unindent_lines()`? Which ones exactly?

**Comment by Paulb23** [comment]
> So should I change them to unindent_lines()? Which ones exactly?

Yeah I think most of them should be okay apart from the ones under the `/* Should indent inplace. */` heading.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**